1-Draft/RFCNNNN-Make-Terminating-Errors-Terminate-The-Right-Way.md

```diff
@@ -0,0 +1,193 @@
+
+In the second workaround, if the shared script does not use `$ErrorActionPreference = 'Stop'`, a caller can get different behaviour by manipulating their `$ErrorActionPreference`. The caller should not be able to manipulate terminating error behavior in commands that they invoke -- that's up to the command author, and they shouldn't have to use extra handling to make terminating errors terminate.
+
+Now consider this code snippet:
+
+```PowerShell
+New-Module -Name ThisShouldNotImport {
+    $myList = [System.Collections.Generics.List[string]]::new()
+
+    function Test-RFC {
+        [CmdletBinding()]
+        param()
+        'Some logic'
+        1 / 0 # Oops
+        'Some more logic'
+    }
+} | Import-Module
+```
+
+There are several things wrong with this snippet:
+
+1. If you invoke that snippet, the `ThisShouldNotImport` module imports successfully because the terminating error (`[System.Collections.Generics.List[string]]` is not a valid type name) does not actually terminate the loading of the module. This could cause your module to load in an unexpected state, which is a bad idea.
+1. If you loaded your module by invoking a command defined with that module, you won't see the terminating error that was raised during the loading of the module (the terminating error that was raised during the loading of the module is not shown at all in that scenario!), so you could end up with some undesirable behaviour when that command executes even though the loading of the module generated a "terminating" error, and not have a clue why.
+1. The Test-RFC command exported by this module produces a terminating error, yet continues to execute after that error.
+1. If the caller either loads your module or invokes your command inside of a `try` block, they will see different behaviour.
+
+Any execution of code beyond a terminating error should be intentional, not accidental like it is in both of these cases, and it most certainly should not be influenced by whether or not the caller loaded the module or invoked the command inside of a `try` block. Binary modules do not behave this way. Why should script modules be any different?
+
+Now have a look at the same module definition, this time with some extra scaffolding in place to make sure that terminating errors actually terminate:
+
+```PowerShell
+New-Module -Name ThisShouldNotImport {
+    trap {
+        break
+    }
+    $myList = [System.Collections.Generic.List[string]]::new()
+
+    function Test-RFC {
+        [CmdletBinding()]
+        param()
+        $callerEAP = $ErrorActionPreference
+        try {
+            'Some logic'
+            1 / 0 # Oops
+            'Some more logic'
+        } catch {
+            Write-Error -ErrorRecord $_ -ErrorAction $callerEAP
+        }
```

Comment:
The pattern that I use is `try { ... } finally {}`, which is a lot simpler to mindlessly type (because the `finally` block is empty; whereas here the `catch` block has content). Is there a reason this is better? I get that we're lobbying to make this sort of workaround unnecessary, but we might as well the real best workarounds.

Reply:
The pattern shared above is the only way to write advanced functions that:

1. Properly terminate on terminating error. The `try/catch` block handles this part.
2. Respect the callers error action preference, such that if the function is invoked with `-ErrorAction Stop`, it will properly stop execution in the caller's scope. This part happens by capturing and then reusing the callers `$ErrorActionPreference` in the `Write-Error` invocation inside of the `catch` block.

That's why this boilerplate logic has been adopted by many. It's the only way to write advanced functions that truly function like cmdlets when it comes to terminating errors.

Reply:
I see; thank you for explaining. The honoring of the EAP is the key here. My confusion here just underscores the need for this RFC.

---
In reply to: [302061411](https://github.com/PowerShell/PowerShell-RFC/pull/187#discussion_r302061411) [](ancestors = 302061411)